### ISSUE TITLE
Demo slice: tiny quantized model on Apple Silicon

### ISSUE THREAD

**Issue by saeedadeeb103** [Abdullah-Taha9/ui-training#57]
## Why this exists

Main is frontend-only. Inspect / Datasets / Training look like a product, but nothing loads weights, writes a checkpoint, or runs a trainer. This issue is the shortest path to a recordable demo on an M4: edit the structure, freeze some layers, drop some layers, change how the embedder is trained, then actually run on a small quantized model and a small dataset.

This is a vertical cut of EPIC 1 / 5 / 7 / 8 / 9, not a new product surface. Full V1 (IR freeze, hosted demo, mergekit, CUDA QLoRA) stays on those epics.

## Demo target

| Piece | Choice | Why |
|---|---|---|
| Model | `HuggingFaceTB/SmolLM2-135M` (already a fixture). Optional 4-bit MLX convert. | ~270 MB fp16; fits M4 unified memory without bitsandbytes |
| Quant | MLX int4, or fp16 on MPS if we skip quant for v0 | bitsandbytes is CUDA-only and will not run on Apple Silicon |
| Dataset | Bundled sample JSONL already in the Datasets tab (~4–6 rows is too small to learn; use ~50–200 instruction pairs for the recording) | No Hub download during the video |
| Chip | Mac M4, local, no cloud GPU | Matches the machine we have |

Do **not** plan Unsloth / bitsandbytes QLoRA for this slice.

## What the video should show

1. Pick SmolLM2-135M.
2. Freeze early layers; LoRA (or full) on the last few blocks.
3. Drop trailing layers (layer-count control already exists in the popup — it must become a real checkpoint rewrite).
4. Change the embedder treatment: freeze `embed_tokens`, or untie `lm_head` via `modules_to_save`. Replacing the embedding matrix / vocab size is EPIC 13 and is **out of this slice** (it needs weight remapping).
5. Map the small JSONL (prompt/response) and show the assembled example.
6. Start training; live loss; finish one short run.
7. Show the adapter / sliced checkpoint on disk.

## Build list (in order)

- [ ] Minimal FastAPI app: `/health`, load fixture or introspect one local/Hub model, return the tree the UI already knows.
- [ ] Compile the client plan (`frozen` / `lora` / `full` per path) to a real PEFT `LoraConfig` + `requires_grad` list. Verification: the wrapped modules equal the selection (the product rule).
- [ ] Layer drop: rewrite `config.json` `num_hidden_layers` and strip `model.layers.{n}.*` keys from the checkpoint (or MLX equivalent). Duplicate can wait. mergekit can wait.
- [ ] Embedder: freeze vs train `embed_tokens`; optional untie `lm_head`. Do not apply hidden-size / vocab-size edits.
- [ ] Local trainer on MLX or `transformers`+MPS. Subprocess, not in the API process. Stream step/loss over SSE to the existing Training view (swap the mock).
- [ ] Dataset: file path or the bundled JSONL, same mapping UI, no silent empty rows.
- [ ] Preflight: memory estimate vs unified memory; block a run that will not fit.
- [ ] Export the adapter dir + a standalone `train.py` so the recording can cut to a terminal `ls`.

## Out of scope (stay on their epics)

- Full Architecture IR freeze / OpenAPI codegen (EPIC 0 / 2)
- 70B weightless introspector hosted demo (EPIC 1 / 12)
- mergekit duplicate/reorder/interleave (rest of EPIC 7)
- Dimension surgery / new embedder weights (EPIC 13)
- CUDA QLoRA, multi-GPU, Unsloth

## Acceptance

A screen recording on an M4: inspect SmolLM2 → freeze some layers → drop some layers → freeze or untie the embedder → map a small JSONL → train until a real loss curve moves → files on disk. One step of training is enough; the curve must come from the subprocess, not `Math.random()`.

**Comment by saeedadeeb103** [comment]
Started the demo slice against Qwen/Qwen2.5-0.5B-Instruct (already in the fixture corpus, ~494M, fits this 16 GB M4 in fp16 LoRA).

Landed:
- packages/core — plan compile + trailing layer-drop (no torch). A pattern that matches nothing is an error.
- packages/server — /health, POST /v1/compile, POST /v1/runs, SSE events. OpenAPI types in web/src/api/types.gen.ts.
- packages/runner — local subprocess trainer (MPS / CUDA / CPU). Writes adapter/, sliced/, and a standalone train.py.
- Training tab talks to the API. Default model in the picker is Qwen 0.5B. Sample instructions dataset is 46 rows.

To actually train (not just compile):

```
uv sync --extra train
make api
make web
```

First run downloads the Hub checkpoint (~1 GB) into the Hugging Face cache. Then: LoRA on the last blocks, drop trailing layers, freeze or fully train embed_tokens, map the JSONL, Start. Loss has to come from the subprocess.

Still out of this issue: full IR freeze, 70B introspector, mergekit duplicate/reorder, dimension surgery, CUDA QLoRA.
